--- linux-5.3.18-allyes/include/linux/slab.h	2019-12-18 15:06:00.000000000 +0000
+++ linux-5.3.18-test-mitigation/include/linux/slab.h	2020-02-03 21:20:36.813897571 +0000
@@ -771,3 +771,13 @@
 #endif
 
 #endif	/* _LINUX_SLAB_H */
+
+
+
+#ifndef CONFIG_FLEXIBLE_H_
+#define CONFIG_FLEXIBLE_H_
+
+inline bool store_flexible_len(void *addr, size_t flexible_len, size_t size, size_t offset);
+inline size_t get_flexible_len(void *addr, size_t size, size_t offset);
+
+#endif
